### PR TITLE
detect/sip: add sticky buffers to match headers v5

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket:
https://redmine.openinfosecfoundation.org/issues/6374
https://redmine.openinfosecfoundation.org/issues/7204

Describe changes:
- New sticky buffers have been rewritten in rust
- Existing sticky buffers have been rewritten in rust, except for `sip.method` and `sip.uri` 

Note:
`sip.method` and `sip.uri` require other C functions to be exposed in rust in order to be rewritten.
I can address this in a follow-up PR, I just need some time to review it.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2005

### CODE REVIEW

**Comment by catenacyber** [comment]
CI is red because of https://github.com/OISF/suricata/pull/11622 for your info

**Comment by catenacyber** [comment]
> New sticky buffers have been rewritten in rust

Thanks :-)

> Existing sticky buffers have been rewritten in rust, except for sip.method and sip.uri

Why these 2 exceptions ?

**Comment by glongo** [comment]
> > New sticky buffers have been rewritten in rust
> 
> Thanks :-)
> 
> > Existing sticky buffers have been rewritten in rust, except for sip.method and sip.uri
> 
> Why these 2 exceptions ?

They require other C functions to be exposed in rust, such as `DetectBufferTypeRegisterValidateCallback`, `DetectBufferTypeRegisterValidateCallback` and `DetectUrilenValidateContent`.

File: rust/src/sip/detect.rs [hunk 1]
```diff
@@ -80,88 +126,662 @@ pub unsafe extern "C" fn rs_sip_tx_get_protocol(
...
+unsafe extern "C" fn sip_from_hdr_get_data(
+    tx: *const c_void, flow_flags: u8, local_id: u32, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    sip_get_header_value(
+        tx,
+        local_id,
+        flow_flags,
+        buffer,
+        buffer_len,
+        "From\0".as_ptr() as *const c_char,
```
Comment (catenacyber) [inline rust/src/sip/detect.rs:376]: Cannot we use a rust string or slice instead of going to a c_char pointer ?

File: rust/src/sip/parser.rs [hunk 2]
```diff
@@ -292,7 +311,7 @@ mod tests {
         assert_eq!(req.method, "REGISTER");
         assert_eq!(req.path, "sip:sip.cybercity.dk");
         assert_eq!(req.version, "SIP/2.0");
-        assert_eq!(req.headers["Content-Length"], "0");
+        assert_eq!(req.headers["Content-Length"].first().unwrap(), "0");
```
Comment (catenacyber) [inline rust/src/sip/parser.rs:315]: Could we have a test with a header having multiple values ?

**Comment by catenacyber** [comment]
Looks really nice :-)

Could you rebase it to get green CI ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
I love this.

Just some nits, and mostly a rebase to get green CI

File: rust/src/sip/detect.rs (hunk 1 above)
Comment (glongo) [reply rust/src/sip/detect.rs:376]: The compiler complains about that:
```extern fn uses type `str`, which is not FFI-safe```

Same for string.

**Comment by glongo** [comment]
Replaced with https://github.com/OISF/suricata/pull/11672

**Pull request closed without merging**